src/mono/mono/mini/interp/jiterpreter.c

```diff
@@ -0,0 +1,1082 @@
+		case MINT_LEAVE:
+		case MINT_LEAVE_S:
+			if (*inside_branch_block)
+				return TRACE_CONTINUE;
+
+			return TRACE_ABORT;
+
+		case MINT_LEAVE_CHECK:
+		case MINT_LEAVE_S_CHECK:
+			return TRACE_ABORT;
+
+		case MINT_CALL_HANDLER:
+		case MINT_CALL_HANDLER_S:
+		case MINT_ENDFINALLY:
+		case MINT_RETHROW:
+		case MINT_MONO_RETHROW:
+		case MINT_PROF_EXIT:
+		case MINT_PROF_EXIT_VOID:
+		case MINT_SAFEPOINT:
+			return TRACE_ABORT;
+
+		default:
+		if (
+			// branches
+			// FIXME: some of these abort traces because the trace compiler doesn't
+			//  implement them, but they are rare
+			(opcode >= MINT_BRFALSE_I4) &&
+			(opcode <= MINT_BLT_UN_I8_IMM_SP)
+		) {
+			*inside_branch_block = TRUE;
+			return TRACE_CONTINUE;
+		}
+		else if (
+			// calls
+			// FIXME: many of these abort traces unconditionally because the trace
+			//  compiler doesn't implement them, but that's fixable
+			(opcode >= MINT_CALL) &&
+			(opcode <= MINT_CALLI_NAT_FAST)
+			// (opcode <= MINT_JIT_CALL2)
+		)
+			return *inside_branch_block ? TRACE_CONTINUE : TRACE_ABORT;
+		else if (
+			// returns
+			(opcode >= MINT_RET) &&
+			(opcode <= MINT_RET_U2)
+		)
+			return *inside_branch_block ? TRACE_CONTINUE : TRACE_ABORT;
+		else if (
+			(opcode >= MINT_LDC_I4_M1) &&
+			(opcode <= MINT_LDC_R8)
+		)
+			return TRACE_CONTINUE;
+		else if (
+			(opcode >= MINT_MOV_SRC_OFF) &&
+			(opcode <= MINT_MOV_8_4)
+		)
+			return TRACE_CONTINUE;
+		else if (
+			// binops
+			(opcode >= MINT_ADD_I4) &&
+			(opcode <= MINT_CLT_UN_R8)
+		)
+			return TRACE_CONTINUE;
+		else if (
+			// unops and some superinsns
+			// fixme: a lot of these aren't actually implemented. but they're also uncommon
+			(opcode >= MINT_ADD1_I4) &&
+			(opcode <= MINT_SHR_I8_IMM)
+		)
+			return TRACE_CONTINUE;
+		else if (
+			// math intrinsics
+			(opcode >= MINT_ASIN) &&
+			(opcode <= MINT_MAXF)
+		)
+			return TRACE_CONTINUE;
+		else if (
+			// field operations
+			// the trace compiler currently implements most, but not all of these
+			(opcode >= MINT_LDFLD_I1) &&
+			(opcode <= MINT_LDTSFLDA)
+		)
+			return TRACE_CONTINUE;
+		else if (
+			// indirect operations
+			// there are also a few of these not implemented by the trace compiler yet
+			(opcode >= MINT_LDLOCA_S) &&
+			(opcode <= MINT_STIND_OFFSET_IMM_I8)
+		)
+			return TRACE_CONTINUE;
+		else if (
+			// array operations
+			// some of these like the _I ones aren't implemented yet but are rare
+			(opcode >= MINT_LDELEM_I) &&
+			(opcode <= MINT_GETITEM_LOCALSPAN)
+		)
+			return TRACE_CONTINUE;
+		else
+			return TRACE_ABORT;
+	}
+}
+
+static gboolean
+should_generate_trace_here (InterpBasicBlock *bb, InterpInst *last_ins) {
+	int current_trace_length = 0;
+	// A preceding trace may have been in a branch block, but we only care whether the current
+	//  trace will have a branch block opened, because that determines whether calls and branches
+	//  will unconditionally abort the trace or not.
+	gboolean inside_branch_block = FALSE;
```

Comment:
This serves no purpose currently. Once we hit a branching opcode the bblock ends and we stop the iteration. If you want to continue probing the trace after a branch you should continue with `bb->next_bb` and start with `first_ins` from there.

Reply:
You're right, this used to do something but it doesn't anymore. I'll figure out whether it's dead code to be removed or whether I need to fix it, since the heuristic is working pretty well as-is.